### PR TITLE
HashSet<T> constructor with capacity

### PR DESCRIPTION
**Pull request by MarkusSintonen**
Issue #382
-Added constructor with capacity parameter to HashSet<T>
-Added two unit tests for HashSet<T> capacity constructor

### CODE REVIEW

**Comment by svick** [comment]
I take it this is the implementation for #382?

**Comment by MarkusSintonen** [comment]
Yes, I updated the comment

**Comment by mellinoe** [comment]
Note: this is the error in the CI build:

Generic\HashSet\HashSet_ConstructorTests.cs(291,36): error CS1502: The best overloaded method match for 'System.Collections.Generic.HashSet<int>.HashSet(System.Collections.Generic.IEnumerable<int>)' has some invalid arguments

This is expected because this new method hasn't actually been added to the contract assembly we build internally, which is what the test compiles against.

File: src/System.Collections/tests/Generic/HashSet/HashSet_ConstructorTests.cs [hunk 1]
```diff
@@ -281,5 +281,27 @@ public static void HashSetConstructor_IEn_IEq_Comparer_NotIEquatable()
         }
 
         #endregion
+
+        #region Constructor_Capacity
+
+        //Test 1: Check that EqualityComparer<T>.Default is used
```
Comment (KrzysztofCwalina) [inline src/System.Collections/tests/Generic/HashSet/HashSet_ConstructorTests.cs]: How does it verify that the default comparer is used?

File: src/System.Collections/tests/Generic/HashSet/HashSet_ConstructorTests.cs [hunk 2]
```diff
@@ -281,5 +281,27 @@ public static void HashSetConstructor_IEn_IEq_Comparer_NotIEquatable()
...
+        public static void HashSetConstructor_Capacity_Negative()
+        {
+            HashSet<int> hashSet = new HashSet<int>(13);
+            int[] items = new int[0];
+            HashSetTestSupport<int> driver = new HashSetTestSupport<int>(hashSet, s_intGenerator, items);
+            driver.VerifyHashSet_NegativeTests();
+        }
+
+        #endregion
     }
```
Comment (KrzysztofCwalina) [inline src/System.Collections/tests/Generic/HashSet/HashSet_ConstructorTests.cs]: Should there be tests for the ctor that takes capacity and comparer?

**Comment by stephentoub** [comment]
@mellinoe, the way we currently build our test projects for partial facades makes it impossible to add tests for new APIs, since they're not part of the contract but we build the tests against the contract:
https://github.com/MarkusSintonen/corefx/blob/future/src/System.Collections/tests/System.Collections.Tests.csproj#L18
Do we have a plan for solving this?

**Comment by mellinoe** [comment]
Other than bringing out our contract generation code to GitHub, I don't know that we have a plan for this necessarily. Traditionally we've always built our tests against the contract, so a change like this would force us to update the contract internally as well. Unfortunately it's not really possible for people adding new APIs to do that until the contract generation stuff is out here.

Technically it should be possible to compile these tests against the partial facade itself, but that will require also compiling against mscorlib.dll, as well as potentially other partial facades that are used by the library. For example, System.Diagnostics.Debug's tests are compiled against the library project file, because it uses internal types, and System.Diagnostics.Debug is a partial facade.

The basic process is to add this to project.json to pull in mscorlib:

 "Microsoft.DotNet.CoreCLR": "1.0.0-prerelease",

Additionally you need a few project properties to force-resolve mscorlib.dll from that assembly if you aren't a Partial Facade project:

    <NuGetTargetFrameworkMoniker>DNXCore,Version=v5.0</NuGetTargetFrameworkMoniker>
    <IgnoreArchitectureMismatches>true</IgnoreArchitectureMismatches>
    <PostFilterNugetReferences>true</PostFilterNugetReferences>

**Comment by stephentoub** [comment]
> Technically it should be possible to compile these tests against the partial facade

Separate from (and before) this change, should we update our projects as you suggest? Or is the stance "no new APIs in partial facades" for now until the rolling issues are sorted?

**Comment by MarkusSintonen** [comment]
I wondered why the test compilation failed on my machine. Well at least it wasn't a problem with my own setup. I think it would be valuable to get the full tooling available since this would allow test coverage for future brach API additions.

File: src/System.Collections/tests/Generic/HashSet/HashSet_ConstructorTests.cs (hunk 2 above)
Comment (MarkusSintonen) [inline src/System.Collections/tests/Generic/HashSet/HashSet_ConstructorTests.cs]: More tests should be done once the tooling problems are resolved related to API changes and unit testing.

File: src/System.Collections/src/System/Collections/Generic/HashSet.cs [hunk 3]
```diff
@@ -131,6 +125,30 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
             }
         }
 
+        public HashSet(int capacity, IEqualityComparer<T> comparer)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException("capacity");
```
Comment (jasonwilliams200OK) [inline src/System.Collections/src/System/Collections/Generic/HashSet.cs]: Using C# 6, we can say:

```c#
throw new ArgumentOutOfRangeException(nameof(capacity)); 
```

File: src/System.Collections/src/System/Collections/Generic/HashSet.cs [hunk 4]
```diff
@@ -131,6 +125,30 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
...
+                comparer = EqualityComparer<T>.Default;
+            }
+
+            _comparer = comparer;
+            _lastIndex = 0;
+            _count = 0;
+            _freeList = -1;
+            _version = 0;
+
+            if (capacity > 0)
```
Comment (jasonwilliams200OK) [inline src/System.Collections/src/System/Collections/Generic/HashSet.cs]: This condition is not required?

File: src/System.Collections/src/System/Collections/Generic/HashSet.cs (hunk 3 above)
Comment (akoeplinger) [inline src/System.Collections/src/System/Collections/Generic/HashSet.cs]: @jasonwilliams200OK corefx doesn't use C#6 yet.

**Comment by stephentoub** [comment]
@MarkusSintonen, it looks like you may have accidentally pushed your concurrent bag changes to this PR. Would you be able to fix that? We're still blocked in taking this change on the infrastructure issues around tests, but hopefully that will be stored out soon and then this can be merged. Thanks!

**Comment by stephentoub** [comment]
@mellinoe, I tried your suggested workaround, but it then fails to compile due to lots of duplicated types between mscorlib and the various contracts.  Could you see if there's an easy way to unblock this?

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@MarkusSintonen, sorry for the delay on this.  Let's go ahead on this without the test for the new ctor, and when we have the infrastructure in place, we can then add the test.  Can you squash this down to a single commit? Thanks!

**Comment by stephentoub** [comment]
Replaced by #2862.

**Pull request closed without merging**